### PR TITLE
Refactor TARIC importer - Rebuilding db process

### PR DESCRIPTION
**Pull request by theharq**
#### What this PR does:

This PR refactors the `System rebuild procedure` described [here](https://github.com/alphagov/trade-tariff-backend/wiki/System-rebuild-procedure) which is not working in the current state of the application.

It should take around 13 hours but I had not been able to complete this process. 

( Test made on Macbook Pro 15" 2015)

With this PR:

`bundle exec rake importer:taric:import   3614.28s user 173.00s system 70% cpu 1:29:34.45 total`

#### Notes

I have made different changes in the code in order to have a better performace to get the rebuilding process working again.

- Replace `Nokogiri` with `ox` gem and add XmlParser module to generate a ruby hash from each XML node
- Remove unnecesary modules
- Extract logic from `record_processor.rb` and `record.rb` related to mutate hash from xml node (SRP)
- Replace `begin/recue` blocks for control block
- Reduce object allocations

###### Optimize XML parser

Comparison of time to parse the `OBEXTACTEN.xml` seed file (80 mb) and generated the expected hash from each node:

```ruby
# OBEXTACTEN.xml
# ruby nokogiri.rb  143.88s user 2.28s system 98% cpu 2:28.67 total

# OBEXTACTEN.xml with ActiveSupport::XmlMini.backend = :NokogiriSAX
# ruby nokogiri.rb  37.14s user 0.66s system 98% cpu 38.448 total

# OBEXTACTEN.xml 
# ruby ox.rb  5.30s user 0.11s system 99% cpu 5.418 total
```

###### Optimize persisting process

After the xml parsing optimization, the next step is to optimize the persisting process, so I've used the 
I used the `ruby-prof` gem to find bottlenecks. Helper I've created: [here](https://gist.github.com/theharq/0bb7ff4b30dcc230e711)

###### Profiler stack when persisting data.

Before optimization:
original

After optimization:
optimized

##### Why I'm avoiding validations on the rebuilding process

First I can always check the data after persisting the whole xml file and delete/mark the records that are not valid. 

Second apparently wrapping the query in a transaction is very slow in `sequel`, and we are not doing any validations from the models, so I want to disable them during this process.
but the `create` [implementation](https://github.com/jeremyevans/sequel/blob/master/lib/sequel/model/base.rb#L149-L151) doesn't allow to receive options, so I've changed this.

Now that we are not validating this record, I want to save the cost of select statement after the insert, which is made [here](https://github.com/jeremyevans/sequel/blob/1e9137167faade311f2ee255aba775beecd26cd5/lib/sequel/model/base.rb#L1948-L1951)

Finally I remove the validate [call](https://github.com/alphagov/trade-tariff-backend/blob/master/lib/taric_importer.rb#L38) because this calls the custom conformance validators which are very expensive to call in this process for each record.
And we know that some records will fail this validation process, for example:

```ruby
# record
<AdditionalCodeType @values={:additional_code_type_id=>"7", :validity_start_date=>1970-01-01 00:00:00 UTC, :validity_end_date=>nil, :application_code=>"3", :meursing_table_plan_id=>"01", :national=>nil, :operation=>"C", :operation_date=>nil}>
# conformance_errors:
{:CT3=>["The Meursing Table plan must exist", "The Meursing Table plan must exist", "The Meursing Table plan must exist", "The Meursing Table plan must exist"]}
```

This previous record from the xml file have an attribute `meursing_table_plan_id` which references another model that it doesn't exists yet.

When I refer to "this process" I'm talking about parsing the 6 GB xml file which is an expensive operation.

I'm thinking to make these validations optional so when calling the `rake task` we can disable validations so the rebuilding process of big xml files will be faster and we can have the old way ( with the new xml parser )
and the new way which is focused on performance.

### CODE REVIEW

File: lib/tasks/tariff_importer.rake [hunk 1]
```diff
@@ -23,8 +23,8 @@ namespace :importer do
         # We will be fetching updates from Taric and modifying primary keys
         # so unrestrict it for all models.
         Sequel::Model.descendants.each(&:unrestrict_primary_key)
-
-        TaricImporter.new(ENV["TARGET"], ENV["DATE"]).import
+        Sequel::Model.plugin :skip_create_refresh
```
Comment (matthewford) [inline lib/tasks/tariff_importer.rake]: Nice: could we link to the documentation http://sequel.jeremyevans.net/rdoc-plugins/classes/Sequel/Plugins/SkipCreateRefresh.html

File: spec/unit/tariff_importer/logger_spec.rb [hunk 2]
```diff
@@ -57,7 +57,8 @@
       TaricImporter.new(valid_file).import
     }
 
-    it 'logs an info event' do
+    xit 'logs an info event' do
```
Comment (matthewford) [inline spec/unit/tariff_importer/logger_spec.rb]: we can make it optional for debugging

**Comment by matthewford** [comment]
Happy about this overall - we can assume that the system is valid on the date of the seed files and yes we can run a validation over all the data anyway.

**Comment by matthewford** [comment]
One thing to note: transactions, I think this was so that we don't leave the db in an inconsistant state - should a daily sync fail we should rollback all changes. 

However as the tariff is a point in time db, we can also achieve this by just deleting all data from that import date (effectively what we do when we do a rollback). Basically we want to make sure the Last updated date published is correct, and it the sync fails we dont display half synced data.

**Comment by theharq** [comment]
PR updated, no pending tests, many commits squashed to more a little more meaningful commit messages

**Pull request merged**